### ISSUE TITLE
🌛 Investigate failing `face.new()` when receiving meshes from Blender

### ISSUE THREAD

**Issue by izzylys** [specklesystems/speckle-blender#35]
When _receiving_ complex meshes _originating from blender_, we often run into errors with `faces.new()` regarding either trying to create a face that already exists or "found the same (BMVert) used multiple times". I'm try catching this for now and just continuing if a face fails, but it is something to investigate as it prob means there is an issue somewhere in the mesh `to_speckle` conversions.

**Comment by AlanRynne** [comment]
Ohh!! Nice find! It may be that the mesh `ToSpeckle` conversion does not respect unique vertices in the mesh, and sends several vertices with the same coords as different points on the vertices list. 🤔

I am aware in Rhino/Gh we do remove duplicates before sending to reduce size. This is called `Weld` but in blender it may be something wildly different.

**Comment by izzylys** [comment]
ah thanks for the tip dude!! I will have a look to see if there is a similar command in blender and if it solves this issue ✨

**Comment by tsvilans** [comment]
Adding verts with the same coordinates in Blender should not be an issue. It is often used for making split meshes ("unwelded", in Rhino lingo). Do you have a sample stream ID that I can use to test this? Will try to take some time to have a look...

**Comment by izzylys** [comment]
hey tom! sure thing - [this is the stream](https://speckle.xyz/streams/646cab6494/commits/76aa5660d9) I've been testing with, but I've had these errors receiving most complex streams that were created within blender - particularly ifc files imported with blender bim

**Comment by tsvilans** [comment]
Great, thanks @izzylys I'll check it out when I get a moment.

`ToSpeckle` shouldn't be making any new Blender data, though... so it's a bit suspicious that you're getting Blender mesh problems when _exporting_ meshes...

**Comment by izzylys** [comment]
no probs, thanks for having a look!

this issue is actually regarding _receiving_ meshes. however, since this only happens with streams originating from blender, it is likely due to something not being quite right when converting `ToSpeckle` in the first place

**Comment by tsvilans** [comment]
Aha, ok gotcha. Mysterious... I'll let you know what I find...

**Comment by Mengesh** [comment]
Hi everybody! I stumbled upon this nice project and this bug catched my eye so I tried to trace the issue.

I analyzed a few meshes from the ifc file separately and I detected two types of irregularities concerning blender objects.

1. **Sometimes triangular faces has repeating vertex index.** That means they are actually edges, so I decided to skip them.
 
2. **Sometimes face vertices does not respect winding order.** When this occurs, the error `faces.new(verts): face already exists` appears. I solved this by using `bmesh.ops.contextual_create` instead of `faces.new`, but it's waaaay slower.

Now I have no errors, but it loads 706 objects.

**Comment by izzylys** [comment]
Hey @Mengesh, thanks a lot for looking into this! Exciting that you were able to get to the bottom of it.

Re 1, that seems like a totally sensible solution to this scenario.

Re 2, that's an interesting find! I'm not familiar with the differences between `contextual_create` and `faces.new` and am not sure what the pros and cons are. Perhaps @tsvilans has some insight into why we went with `faces.new`? However, I am hesitant to make the swap if the performance cost is very high. Would be interested in benchmarks for the two methods. 

Side note that there are some serialisation improvements that will greatly reduce the speed of sending / receiving streams coming in specklesystems/specklepy#118. I'd like to continue the upwards trajectory in performance if possible 😅! 

Anyway, I really appreciate your looking into this! It's great to now know what the source of the errors are so we can be confident nothing is being missed during conversion. Since we are already skipping this error at the moment, I don't think any code changes are necessary unless you would like to create a PR for 1 (which I'd be happy to accept) or if we'd like to further discuss / investigate 2.

**Comment by tsvilans** [comment]
Any `ops` functions will be slower; calling `faces.new` is much faster and low-level.
My concern would be that if you're defining a face with the same verts, just with a different winding order, you anyways have a duplicate face, just with a flipped normal. Maybe try wrapping `face.new` in a `try/except` pair and pass on the ones that it doesn't like. Also not optimal, and most likely slower, but might circumvent the problem for now...

**Comment by Mengesh** [comment]
I was wrong. This time I investigated `IfcDoor/TU DF 1 - Stahl-Eckzarge Falztüre:DL - 800 x 2000:2699 117/211261` loaded from the ifc file. I intercepted the blender object before it is sent to speckle and print out all the 4 vertex faces. There are two duplicates:
```
...
891, 2, 3, 11
...
2, 3, 11, 891
...
```
and
```
...
848, 847, 832, 4
...
4, 848, 847, 832
...
```

When I load it from speckle to blender it gives the error `faces.new(verts): face already exists` twice which actually means that those faces already exist :facepalm: What a find!

So I believe the problem is likely inside the ifc file or less likely inside blender, but not speckle. And it's actually safe to skip those faces, like you already do.
